torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -1962,6 +1962,27 @@ std::tuple<Tensor, Tensor> atan2_backward(const Tensor& grad, const Tensor& self
             output_mask[1] ? grad * -self * recip : Tensor() };
 }
 
+std::tuple<Tensor, Tensor> attn_backward(Tensor grad_output, Tensor grad_attn,
+                                         const Tensor& q, const Tensor& k, const Tensor& v,
+                                         const Tensor& attn) {
+  // u3 in the forward pass is attn.matmul(v), aka output. We use a
+  // different name because grad_output is already present as the
+  // gradient of the output result.
+  Tensor grad_u3 = grad_output.defined()
+    ? grad_output.matmul(v.t())
+    : at::zeros({attn.size(0), v.size(0)});  // match dimensions above
```

Comment:
Could you do better than materializing a Tensor full of zeros here? (with the same idea that undefined Tensor is the same as full of zeros for gradients).

Reply:
i can. i went a little overboard here with doing things in-place where allowed. I'm not sure why I can't make the final transform in-place. Any ideas?